core/chain_indexer.go

```diff
@@ -473,3 +492,123 @@ func (c *ChainIndexer) removeSectionHead(section uint64) {
+	utxos := block.QiTransactions()
+
+	for _, tx := range utxos {
+		for i, in := range tx.TxIn() {
+			outpoint := in.PreviousOutPoint
+
+			// Coinbase
+			if i == 0 {
+				continue
+			}
+			addr := crypto.PubkeyBytesToAddress(in.PubKey, config.Location).Hex()
+			outpointsForAddress := addressOutpoints[addr]
+
+			for i := 0; i < len(outpointsForAddress); i++ {
+				if &outpointsForAddress[i].TxHash == &outpoint.TxHash && outpointsForAddress[i].Index == outpoint.Index {
+					outpointsForAddress = append(outpointsForAddress[:i], outpointsForAddress[i+1:]...)
+					addressOutpoints[addr] = outpointsForAddress
+
+					break
+				}
+			}
+		}
+
+		for i, out := range tx.TxOut() {
+
+			address := out.Address
+			outpoint := types.OutPoint{
+				TxHash: tx.Hash(),
+				Index:  uint16(i),
+			}
+
+			addr := common.BytesToAddress(address, config.Location).Hex()
+
+			outpointAndDenom := &types.OutpointAndDenomination{
+				TxHash:       outpoint.TxHash,
+				Index:        outpoint.Index,
+				Denomination: out.Denomination,
+			}
+
+			if existingOutpoints, ok := addressOutpoints[addr]; ok {
+				addressOutpoints[addr] = append(existingOutpoints, outpointAndDenom)
+			} else {
+				addressOutpoints[addr] = []*types.OutpointAndDenomination{outpointAndDenom}
+			}
+		}
+	}
+	err := rawdb.WriteAddressOutpoints(c.chainDb, addressOutpoints)
+	if err != nil {
+		panic(err)
+	}
+
+}
+func (c *ChainIndexer) reorgUtxoIndexer(headers []*types.Header, addressOutpoints map[string][]*types.OutpointAndDenomination, nodeCtx int, config params.ChainConfig) error {
+
+	for _, header := range headers {
+		block := rawdb.ReadBlock(c.chainDb, header.Hash(), header.NumberU64(nodeCtx), config.Location)
+
+		for _, tx := range block.QiTransactions() {
+			for i, out := range tx.TxOut() {
+
+				address := out.Address
+
+				outpoint := types.OutpointAndDenomination{
+					TxHash:       tx.Hash(),
+					Index:        uint16(i),
+					Denomination: out.Denomination,
+				}
+
+				addr := common.BytesToAddress(address, config.Location).Hex()
+				outpointsForAddress := addressOutpoints[addr]
+
+				for i := 0; i < len(outpointsForAddress); i++ {
+					if outpointsForAddress[i] == &outpoint {
+						outpointsForAddress = append(outpointsForAddress[:i], outpointsForAddress[i+1:]...)
+						addressOutpoints[addr] = outpointsForAddress
+
+						break
+					}
+				}
+			}
+
+			for i, in := range tx.TxIn() {
+				outpoint := in.PreviousOutPoint
+
+				// Coinbase
+				if i == 0 {
+					continue
+				}
+				address := crypto.PubkeyBytesToAddress(in.PubKey, config.Location).Hex()
+
+				parent := rawdb.ReadHeader(c.chainDb, block.ParentHash(nodeCtx), uint64(nodeCtx))
+
+				state, err := c.StateAt(parent.EVMRoot(), parent.UTXORoot())
+				if err != nil {
+					return err
+				}
+
+				entry := state.GetUTXO(outpoint.TxHash, outpoint.Index)
+				if entry == nil {
+					return errors.New("failed to get UTXO for address")
```

Comment:
What if this entry is no longer valid because it was created in a block that was removed from the chain? If you stop the re-org in the middle, your indexer has an invalid state

Reply:
We should still have the state if the block is not canonical, correct?